src/Compilers/Core/Portable/DiagnosticAnalyzer/CachingSemanticModelProvider.cs

```diff
@@ -0,0 +1,50 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    /// <summary>
+    /// Provider that caches semantic models for requested trees, with a strong reference to the model.
+    /// Clients using this provider are responsible for maintaining the lifetime of the entries in this cache,
+    /// and should invoke <see cref="EnsureCachedSemanticModelRemoved(SyntaxTree, Compilation)"/> to clear entries when appropriate.
+    /// For example, <see cref="CompilationWithAnalyzers"/> uses this provider to ensure that semantic model instances
+    /// are shared between the compiler and analyzers for improved analyzer execution performance. The underlying
+    /// <see cref="AnalyzerDriver"/> executing analyzers clears entries in the cache whenever a <see cref="CompilationUnitCompletedEvent"/>
+    /// has been processed, indicating all relevant analyzers have executed on the corresponding syntax tree for the event.
```

Comment:
Is there a corresponding event for "All trees have finished processing"? That would allow us to ensure that the entire cache is flushed at that point as an invariant.

Reply:
Yes, there is a [CompilationCompleted](http://sourceroslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/CompilationCompletedEvent.cs,a37b5031cc734702,references), but I don't think we can assert the invariant as we could have inefficiently written analyzers that can invoke `compilation.GetSemanticModel` on random trees in the compilation at any time. This would be something that needs the analyzer to be fixed to avoid holding onto the model for the lifetime of compilation, but I am wondering if we should convert the strong reference via a Dictionary here instead be a weak reference via ConditionalWeakTable. I had measured performance for both these data structures on the compiler benchmark and did not see any noticeable difference so that would also be a feasible approach.

Reply:
Hmm. A concern with that type of analyzer: such an analyzer could request a model for a tree that has already been processed, and since the compilation is aware of the cache, it'll be added. What then clears that model from the cache? Unless I'm misunderstanding something, this is likely a memory leak. We could solve that with a CWT, but what would be the key of the CWT in this instance? The syntax tree? Or a `ConditionalWeakTable<Compilation, ConditionalWeakTable<SyntaxTree, SemanticModel>>`? It really does feel like the best solution would be, on throwing away a compilation, we throw away any cached models associated with that compilation.

Reply:
It feels like we either need to have the CWT of CWT solution, or more strongly tie a provider to only being used for a single compilation, otherwise we'll run into memory leak issues.

Reply:
> It feels like we either need to have the CWT of CWT solution, or more strongly tie a provider to only being used for a single compilation, otherwise we'll run into memory leak issues.

Agreed. I like the latter approach as we already take a similar approach for event queue, and we already ensure that we use a new caching semantic provider for each CompilationWithAnalyzers instance, i.e. it is tied to a specific compilation instance. I will make that change.

Reply:
I put a bit more thought into this, and tying a provider instance to compilation likely prevents it from ever being made public, similar to event queue on the compilation. I went with a hybrid route of `CWT<Compilation, ConcurrentDictionary<Tree, Model>>`, with the host appropriately clearing per-tree entries in the dictionary value on `CompilationUnitCompletedEvent` and clearing the CWT entry on `CompilationCompletedEvent`. I will measure and report perf numbers on the benchmark with this change.

Reply:
I validated the perf numbers on the benchmark with the last commit and numbers seem similar to the ones with original implementation:

```
|                      Method |    Mean |   Error |  StdDev |  Median |     Min |     Max |       Gen 0 |      Gen 1 |     Gen 2 | Allocated |
|---------------------------- |--------:|--------:|--------:|--------:|--------:|--------:|------------:|-----------:|----------:|----------:|
| GetDiagnosticsWithAnalyzers | **12.63 s** | 0.066 s | 0.055 s | 12.61 s | 12.56 s | 12.76 s | 208000.0000 | 65000.0000 | 1000.0000 |   **1.23 GB** |
```